### PR TITLE
init upgradehandler (pre-pivot) and cleanup

### PR DESCRIPTION
**Pull request by pixelsoccupied**
- init code for upgrade prepivot
  - Call and backup and extra-manifest as needed and moved some of the functions to upgradeHandler
  - No requeue-ing when an error is caught
- an error msg for when ibu name is not what's expected
- CRs to quickly deploy minIO and other support CRs+doc to work with backup in HACKING.md
- include RBAC for deletebackuprequests

### CODE REVIEW

**Comment by pixelsoccupied** [comment]
/hold 

plugging in ExportRestoresToDir

File: HACKING.md [hunk 1]
```diff
@@ -140,4 +144,133 @@ perform some checks in the seed SNO cluster.
 
 ```shell
 skopeo copy docker://quay.io/edge-infrastructure/recert docker://${LOCAL_REGISTRY}/edge-infrastructure/recert
+````
+
+## Setup dev backup steps
+
+### minio + oadp oprator
```
Comment (Missxiaoguo) [inline HACKING.md:151]: This can be easy as running this command. Perhaps provide this option as well to bring up minio in a container.
`podman run --name minio -d -p 9000:9000 -p 9001:9001 quay.io/minio/minio server /data --console-address ":9001"
`

File: controllers/upgrade_handlers.go [hunk 2]
```diff
@@ -21,36 +21,101 @@ import (
...
+// handleUpgrade orchestrate main upgrade steps
 func (r *ImageBasedUpgradeReconciler) handleUpgrade(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+	// log and update status to in-progress
+	r.Log.Info("Starting handleUpgrade")
+	utils.SetStatusCondition(&ibu.Status.Conditions,
+		utils.GetInProgressConditionType(lcav1alpha1.Stages.Upgrade),
+		utils.ConditionReasons.InProgress,
+		metav1.ConditionTrue,
+		"Upgrade in-progress",
+		ibu.Generation)
```
Comment (Missxiaoguo) [inline controllers/upgrade_handlers.go:40]: This will override the old status everytime it's re-queued and perform uncessary steps which were previously completed. We should check the current IBU status to determine if move to the next step.

File: controllers/upgrade_handlers.go [hunk 3]
```diff
@@ -21,36 +21,101 @@ import (
...
-		return ctrl.Result{}, err
+		msg := "failed to export Restore CRs"
+		r.Log.Error(err, msg)
+		utils.SetStatusCondition(&ibu.Status.Conditions,
+			utils.GetInProgressConditionType(lcav1alpha1.Stages.Upgrade),
+			utils.ConditionReasons.Failed,
+			metav1.ConditionFalse,
+			fmt.Sprintf("%s: %v", msg, err),
+			ibu.Generation)
+		return requeueWithError(err)
```
Comment (Missxiaoguo) [inline controllers/upgrade_handlers.go:92]: We want to invoke ExportOadpConfigurationToDir() to export the DataProtectionApplication CR and secrets as well.

File: controllers/upgrade_handlers.go [hunk 4]
```diff
@@ -21,36 +21,101 @@ import (
...
 
-	// TODO: Pre-pivot steps
-	//
-	// TODO: Call r.BackupRestore.ExportRestoresToDir to write restore CRs to the new stateroot.
-	//       If there is invalid type of error returned, we should update the ibu status and
-	//       requeue with short interval.
-	//
-	err := r.ExtraManifest.ExportExtraManifestToDir(ctx, ibu.Spec.ExtraManifests, stateRootRepo)
+	r.Log.Info("Writing extra-manifests into new stateroot")
+	if err = r.ExtraManifest.ExportExtraManifestToDir(ctx, ibu.Spec.ExtraManifests, stateRootRepo); err != nil {
```
Comment (Missxiaoguo) [inline controllers/upgrade_handlers.go:69]: As we have done the backups previously, could we then export the restore CRs and OADP configurations prior to exporting extramanifests

File: controllers/ibu_controller.go [hunk 5]
```diff
@@ -99,40 +103,38 @@ func requeueWithCustomInterval(interval time.Duration) ctrl.Result {
...
-	}()
 
-	nextReconcile = doNotRequeue()
+	// default to not requeue
+	nextReconcile := doNotRequeue()
 
 	if req.Name != utils.IBUName {
-		return
+		msg := fmt.Errorf("invalid IBU name, it must be named '%s'", utils.IBUName)
+		return doNotRequeue(), msg
```
Comment (jc-rh) [inline controllers/ibu_controller.go:114]: If you return msg as an error, k8s will retry by itself.

File: controllers/ibu_controller.go [hunk 6]
```diff
@@ -99,40 +103,38 @@ func requeueWithCustomInterval(interval time.Duration) ctrl.Result {
...
 	r.Log.Info("Loaded IBU", "name", req.NamespacedName, "version", ibu.GetResourceVersion(), "desired stage", ibu.Spec.Stage)
 
+	// main
 	currentInProgressStage := utils.GetCurrentInProgressStage(ibu)
+
 	if currentInProgressStage != "" {
 		nextReconcile, err = r.handleStage(ctx, ibu, currentInProgressStage)
 		if err != nil {
-			return
+			return requeueWithError(err)
```
Comment (jc-rh) [inline controllers/ibu_controller.go:137]: This is just to make it more descriptive? Personally I prefer more concise code with named return vars.

File: controllers/upgrade_handlers.go (hunk 2 above)
Comment (jc-rh) [reply controllers/upgrade_handlers.go:40]: Yes, the stage handlers are not one shot. It can get called repeatedly if waiting for long running background jobs. The initial in progress condition is already set by the main controller when stage transition happens. The handlers can update the reason/message along the way until it reaches the end (status false with either success or failure)

Comment (pixelsoccupied) [reply controllers/upgrade_handlers.go]: Removed it!

File: controllers/upgrade_handlers.go (hunk 3 above)
Comment (pixelsoccupied) [reply controllers/upgrade_handlers.go:99]: Added ExportOadpConfigurationToDir PTAL

File: controllers/ibu_controller.go (hunk 5 above)
Comment (pixelsoccupied) [reply controllers/ibu_controller.go]: Updated to return nil

File: controllers/ibu_controller.go (hunk 6 above)
Comment (pixelsoccupied) [reply controllers/ibu_controller.go:138]: Yeah agreed named func parameter with no return values are cleaner. But it's only really good for small functions and large functions like our reconcile it was getting pretty tricky to follow without being explicit. 

Let me know if you like me revert it no worries!

File: HACKING.md (hunk 1 above)
Comment (pixelsoccupied) [reply HACKING.md:151]: Added doc

File: controllers/upgrade_handlers.go (hunk 4 above)
Comment (pixelsoccupied) [reply controllers/upgrade_handlers.go]: Updated

File: controllers/ibu_controller.go (hunk 6 above)
Comment (jc-rh) [reply controllers/ibu_controller.go:138]: I see that you also removed the defer block. That was the main reason why I did that way for TALM. There were too many returns with all sorts of requeueAfter values and I didn't want to duplicate the logging lines so many times. This one is not too bad yet. Hope it stays that way :)

File: controllers/upgrade_handlers.go [hunk 7]
```diff
@@ -18,39 +18,125 @@ package controllers
...
+
 	"github.com/openshift-kni/lifecycle-agent/controllers/utils"
+	"github.com/openshift-kni/lifecycle-agent/internal/backuprestore"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	ctrl "sigs.k8s.io/controller-runtime"
 )
 
+// handleUpgrade orchestrate main upgrade steps
 func (r *ImageBasedUpgradeReconciler) handleUpgrade(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+	// log and update status to in-progress
```
Comment (Missxiaoguo) [inline controllers/upgrade_handlers.go:34]: The overall flow is not that correct.  Comments for the high level, 
- The handler could be repeatedly called when waiting for long running process or if error happens. It shouldn't start from beginning everytime it's being called. 
- We shouldn't record every single error in the IBU status. Most of them could be just API call issues due to unstable network.

File: controllers/ibu_controller.go [hunk 8]
```diff
@@ -191,6 +199,12 @@ func (r *ImageBasedUpgradeReconciler) handleStage(ctx context.Context, ibu *lcav
 	case lcav1alpha1.Stages.Rollback:
 		nextReconcile, err = r.handleRollback(ctx, ibu)
 	}
+
```
Comment (jc-rh) [inline controllers/ibu_controller.go:202]: If there is any hander error, it gets returned to k8s and any update made here gets discarded. K8s basically calls the reconcile again with the original version of the CR.

File: controllers/upgrade_handlers.go [hunk 9]
```diff
@@ -18,39 +18,125 @@ package controllers
...
-	//
-	err := r.ExtraManifest.ExportExtraManifestToDir(ctx, ibu.Spec.ExtraManifests, stateRootRepo)
+	r.Log.Info("Writing OadpConfiguratio CRs into new stateroot")
+	err = r.BackupRestore.ExportOadpConfigurationToDir(ctx, stateRootRepo, backuprestore.OadpNs)
 	if err != nil {
-		r.Log.Error(err, "Failed to export extra manifests")
-		return ctrl.Result{}, err
+		msg := "failed to export OadpConfiguratio CRs"
+		r.Log.Error(err, msg)
+		utils.SetStatusCondition(&ibu.Status.Conditions,
```
Comment (jc-rh) [inline controllers/upgrade_handlers.go:69]: If we want to declare upgrade failure in the conditions, don't return the err back to k8s.
If it's an unexpected/transient error, don't set the conditions and let k8s retry.

File: controllers/upgrade_handlers.go [hunk 10]
```diff
@@ -18,39 +18,125 @@ package controllers
...
+// handleUpgrade orchestrate main upgrade steps
 func (r *ImageBasedUpgradeReconciler) handleUpgrade(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+	// log and update status to in-progress
+	r.Log.Info("Starting handleUpgrade")
+
+	// Pre-pivot steps
+
+	// backup with OADP
+	r.Log.Info("Starting backup")
+	backupRequeue, backupStatus, err := r.BackupRestore.ReconcileBackup(ctx, ibu.Spec.OADPContent)
```
Comment (jc-rh) [inline controllers/upgrade_handlers.go:41]: I don't see when we will get completed status here unless the spec is empty. @Missxiaoguo should triggerBackup return something if all backup jobs found completed?

File: controllers/upgrade_handlers.go (hunk 9 above)
Comment (jc-rh) [reply controllers/upgrade_handlers.go:69]: Also, if we decide we are done with this stage (no more retry or requeue), both conditions need to be set, whether success or failure.

**Comment by pixelsoccupied** [comment]
/hold 

making some adjustments

File: controllers/upgrade_handlers.go (hunk 10 above)
Comment (Missxiaoguo) [reply controllers/upgrade_handlers.go:41]: Right, the triggerBackup returns BackupCompleted status if all backups are completed.

File: internal/backuprestore/backup.go [hunk 11]
```diff
@@ -423,6 +424,11 @@ func (h *BRHandler) ExportRestoresToDir(ctx context.Context, configMaps []lcav1a
 
 	restores, err := extractRestoreFromConfigmaps(ctx, h.Client, configmaps)
 	if err != nil {
+		// detects validation error with dryrun
+		if !k8serrors.IsNotFound(err) {
+			ErrInvalidCR = err
+			return ErrInvalidCR
+		}
```
Comment (Missxiaoguo) [inline internal/backuprestore/backup.go:431]:
```suggestion
	if err != nil {
	         return err
         }
```
There are other error types other than notfound and invalid.  For particular invalid error type returned by the API call should be k8serrors.IsInvalid(err).  And in the main handlerUpgrade function, you should check the if it's k8serrors.IsInvalid(err)

File: internal/backuprestore/backup.go [hunk 12]
```diff
@@ -41,27 +41,27 @@ import (
...
+		result = ctrl.Result{}
+	)
 	if len(oadpContent) == 0 {
-		status.Status = BackupFailedValidation
-		status.Message = "No oadp configmap is provided."
-		result.RequeueAfter = 1 * time.Minute
-		return
+		status.Status = BackupCompleted
+		status.Message = "No oadp configmap is provided. Skipping backup"
+		return result, status, nil
```
Comment (Missxiaoguo) [inline internal/backuprestore/backup.go:52]: Add an info log here to indicate no oadp configmap is provided.  For this case, the upgrade will continue and we don't need to update the IBU status.

File: controllers/upgrade_handlers.go [hunk 13]
```diff
@@ -18,39 +18,125 @@ package controllers
...
+	if err != nil {
+		r.Log.Error(err, fmt.Sprintf("error msg from backup: %s, will requeue after: %s", backupStatus.Message, backupRequeue.RequeueAfter))
+		return backupRequeue, err
+	}
+	switch backupStatus.Status {
+	case backuprestore.BackupCompleted:
+		// only proceed when BackupCompleted
+		r.Log.Info(fmt.Sprintf("Backup completed with msg %s", backupStatus.Message))
+	default:
+		// backup in progress
```
Comment (Missxiaoguo) [inline controllers/upgrade_handlers.go:51]: Don't forget to include the cases BackupFailed(failed upgrade, no requeue), BackupFailedValidation(requeue)

File: controllers/upgrade_handlers.go (hunk 7 above)
Comment (pixelsoccupied) [reply controllers/upgrade_handlers.go:40]: Updated like we discussed offline

File: controllers/ibu_controller.go (hunk 6 above)
Comment (pixelsoccupied) [reply controllers/ibu_controller.go]: Reverted!

File: controllers/ibu_controller.go (hunk 8 above)
Comment (pixelsoccupied) [reply controllers/ibu_controller.go]: Reverted

File: internal/backuprestore/backup.go (hunk 11 above)
Comment (pixelsoccupied) [reply internal/backuprestore/backup.go]: Updated to use Invalid

File: controllers/upgrade_handlers.go (hunk 9 above)
Comment (pixelsoccupied) [reply controllers/upgrade_handlers.go]: We are not requeue-ing on any errors and i brought back the statues that were previously there

File: internal/backuprestore/backup.go (hunk 12 above)
Comment (pixelsoccupied) [reply internal/backuprestore/backup.go]: no more logging as discussed

File: controllers/upgrade_handlers.go [hunk 14]
```diff
@@ -19,38 +19,134 @@ package controllers
...
+	if err != nil {
+		r.Log.Error(err, "")
+		return doNotRequeue(), nil
+	}
+	if reqBackup {
+		// validate if backup CR is not malformed
+		err := backuprestore.ValidateBackupCRs(ctx, r.BackupRestore, backupCRs)
+		if err != nil {
+			r.Log.Error(err, "")
+			return doNotRequeue(), nil
```
Comment (jc-rh) [inline controllers/upgrade_handlers.go:87]: Are these errors that can be corrected by user? If they are, we would want to requeue with an interval.

Comment (pixelsoccupied) [reply controllers/upgrade_handlers.go:87]: Yeah! But yesterday during we return regardless if can be fixed or not?

File: controllers/upgrade_handlers.go [hunk 15]
```diff
@@ -76,5 +172,167 @@ func (r *ImageBasedUpgradeReconciler) handleUpgrade(ctx context.Context, ibu *lc
 		metav1.ConditionFalse,
 		"Upgrade completed",
 		ibu.Generation)
+
 	return doNotRequeue(), nil
 }
+
+func checkIfBackupRequested(restore *backuprestore.BRHandler, content []lcav1alpha1.ConfigMapRef, ctx context.Context) (bool, []*velerov1.Backup, error) {
```
Comment (jc-rh) [inline controllers/upgrade_handlers.go:179]: Does it make more sense to have these functions as part of the backup plugin code?

Comment (pixelsoccupied) [reply controllers/upgrade_handlers.go]: Yeah! Moved to backup.go

File: controllers/upgrade_handlers.go [hunk 16]
```diff
@@ -62,7 +158,7 @@ func (r *ImageBasedUpgradeReconciler) handleUpgrade(ctx context.Context, ibu *lc
 	// 	 r.Log.Error(err, "Failed to apply extra manifests")
 	//	 return ctrl.Result{}, err
 	// }
-
+	r.Log.Info("Done handleUpgrade")
```
Comment (Missxiaoguo) [inline controllers/upgrade_handlers.go:161]: The IBU status condition for UpgradeCompleted and UpgradeInProgress should be updated to False which indicates the upgrade fails.

File: controllers/upgrade_handlers.go [hunk 17]
```diff
@@ -76,5 +172,146 @@ func (r *ImageBasedUpgradeReconciler) handleUpgrade(ctx context.Context, ibu *lc
...
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+func (r *ImageBasedUpgradeReconciler) CreateNewBackupCr(ctx context.Context, backup *velerov1.Backup) error {
```
Comment (Missxiaoguo) [inline controllers/upgrade_handlers.go:201]: As we talked, this shouldn't be splitted out of the triggerBackup function. If the backup requires ordering, the previous group should be completed prior to applying the second group.

Comment (Missxiaoguo) [reply controllers/upgrade_handlers.go:201]: In addition to that, if the upgrade handler is an one-shot, you will need to periodically check for the resulf of the backup CR.  If it's not an one-shot, you will return requeue interval and check the result in the next intervel.

File: controllers/upgrade_handlers.go (hunk 14 above)
Comment (jc-rh) [reply controllers/upgrade_handlers.go:87]: See Angie's comment too. We either
- keep this upgrade alive by requeuing and maybe updating in progress condition message/reason only, or
- give it up by doNotRequeue and updating the conditions to false.

The only exception is an error that can only be corrected by updating IBU CR spec. In that case, we can return doNotRequeue while keeping the upgrade still in progress.

File: controllers/upgrade_handlers.go [hunk 18]
```diff
@@ -76,5 +172,146 @@ func (r *ImageBasedUpgradeReconciler) handleUpgrade(ctx context.Context, ibu *lc
...
+				return err
+			}
+
+			if existingBackup == nil {
+				// Backup has not been applied. Applying backup
+				err := r.CreateNewBackupCr(ctx, backup)
+				if err != nil {
+					return err
+				}
+			}
```
Comment (pixelsoccupied) [inline controllers/upgrade_handlers.go:195]: call GetBackupCtrlResult from here

File: internal/backuprestore/backup.go [hunk 19]
```diff
@@ -295,17 +121,10 @@ func extractBackupFromConfigmaps(ctx context.Context, c client.Client, configmap
...
 				continue
 			}
 
-			// Create the backup CR in dry-run mode to detect any validation errors in the CR
-			// i.e., missing required fields
-			err = c.Create(ctx, &resource, &client.CreateOptions{DryRun: []string{metav1.DryRunAll}})
-			if err != nil {
-				return nil, err
-			}
-
```
Comment (Missxiaoguo) [inline internal/backuprestore/backup.go:308]: Forget to mention. You need to keep this dryrun here for the "unstructured" resource type. With the CR marshalled into velerov1.backup type, the missing required fields will not be catched as they will be filled with empty string.

File: internal/backuprestore/backup.go [hunk 20]
```diff
@@ -35,218 +34,45 @@ import (
...
-	return
-}
-
-func equivalentBackups(backup1, backup2 *velerov1.Backup) bool {
+func EquivalentBackups(backup1, backup2 *velerov1.Backup) bool {
 	// Compare the specs only
 	return equality.Semantic.DeepEqual(backup1.Spec, backup2.Spec)
 }
 
-func (h *BRHandler) checkVeleroBackupProcessStatus(backup *velerov1.Backup) BackupPhase {
```
Comment (Missxiaoguo) [inline internal/backuprestore/backup.go:216]: Prefer you keep this function as-is. Have the log added in this function

File: controllers/upgrade_handlers.go (hunk 16 above)
Comment (pixelsoccupied) [reply controllers/upgrade_handlers.go:130]: Updated

File: internal/backuprestore/backup.go (hunk 20 above)
Comment (pixelsoccupied) [reply internal/backuprestore/backup.go:216]: Updated

File: internal/backuprestore/backup.go (hunk 19 above)
Comment (pixelsoccupied) [reply internal/backuprestore/backup.go]: Reverted

File: controllers/upgrade_handlers.go (hunk 17 above)
Comment (pixelsoccupied) [reply controllers/upgrade_handlers.go]: Reverted + updated as discussed

File: controllers/upgrade_handlers.go (hunk 14 above)
Comment (pixelsoccupied) [reply controllers/upgrade_handlers.go]: Updated as discussed. Following the more traditional path

File: controllers/upgrade_handlers.go [hunk 21]
```diff
@@ -18,39 +18,104 @@ package controllers
...
 func (r *ImageBasedUpgradeReconciler) handleUpgrade(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+	r.Log.Info("Starting handleUpgrade")
+
+	// backup with OADP
+	r.Log.Info("Backup with Oadp operator")
+	reqBackup, backupCRs, err := r.BackupRestore.CheckIfBackupRequested(ibu.Spec.OADPContent, ctx)
+	if err != nil {
+		if k8serrors.IsInvalid(err) {
+			utils.SetStatusCondition(&ibu.Status.Conditions,
+				utils.GetInProgressConditionType(lcav1alpha1.Stages.Upgrade),
```
Comment (Missxiaoguo) [inline controllers/upgrade_handlers.go:52]: I think the reason should be InProgress

File: controllers/upgrade_handlers.go [hunk 22]
```diff
@@ -18,39 +18,104 @@ package controllers
...
+	if reqBackup {
+		r.Log.Info("Backup requested")
+		ctrlResult, err := r.handleBackup(ctx, backupCRs)
+		if err != nil {
+			switch {
+			case errors.Is(err, BackupFailed):
+				utils.SetStatusCondition(&ibu.Status.Conditions,
+					utils.GetInProgressConditionType(lcav1alpha1.Stages.Upgrade),
+					utils.ConditionReasons.Failed,
+					metav1.ConditionFalse,
```
Comment (Missxiaoguo) [inline controllers/upgrade_handlers.go:70]: The status should be True. And we also need to update the UpgradeCompleted condition.

File: controllers/upgrade_handlers.go [hunk 23]
```diff
@@ -18,39 +18,104 @@ package controllers
...
+					"Upgrade failed",
+					ibu.Generation)
+				return doNotRequeue(), nil
+			case errors.Is(err, BackupValidationFailedShortRequeue):
+				utils.SetStatusCondition(&ibu.Status.Conditions,
+					utils.GetInProgressConditionType(lcav1alpha1.Stages.Upgrade),
+					utils.ConditionReasons.Failed,
```
Comment (Missxiaoguo) [inline controllers/upgrade_handlers.go:77]: I think the reason should be InProgress

File: internal/backuprestore/backup.go [hunk 24]
```diff
@@ -35,185 +34,125 @@ import (
...
+// Constants for backup phase
+const (
+	BackupPending          BackupPhase = "BackupPending"
+	BackupFailedValidation BackupPhase = "BackupFailedValidation"
+	BackupFailed           BackupPhase = "BackupFailed"
+	BackupCompleted        BackupPhase = "BackupCompleted"
+	BackupInProgress       BackupPhase = "BackupInProgress"
+	BackupSkipped          BackupPhase = "BackupSkipped"
+	BackupRequested        BackupPhase = "BackupRequested"
+)
```
Comment (Missxiaoguo) [inline internal/backuprestore/backup.go:61]: As we are not returning a status, could these be removed?

File: controllers/upgrade_handlers.go [hunk 25]
```diff
@@ -76,5 +141,49 @@ func (r *ImageBasedUpgradeReconciler) handleUpgrade(ctx context.Context, ibu *lc
...
+			msg := fmt.Sprintf("Failed backups: %s", strings.Join(backupTracker.FailedBackups, ","))
+			BackupFailed = fmt.Errorf("%s: %s", BackupFailed.Error(), msg)
+			r.Log.Error(BackupFailed, "")
+			return doNotRequeue(), BackupFailed
+		} else if len(backupTracker.ProgressingBackups) != 0 {
+			r.Log.Info(fmt.Sprintf("Inprogress backups: %s", strings.Join(backupTracker.ProgressingBackups, ",")))
+			return requeueWithCustomInterval(2 * time.Second), nil
+		} else if len(backupTracker.FailedValidationBackups) != 0 {
+			msg := fmt.Sprintf("Failed validation backups reported by Oadp: %s. %s", strings.Join(backupTracker.FailedValidationBackups, ","), "Please update the invalid backups.")
+			BackupValidationFailedShortRequeue = fmt.Errorf(msg)
```
Comment (Missxiaoguo) [inline controllers/upgrade_handlers.go:176]: This is defined as gobal variable on the top. Will that be overriden?

File: controllers/upgrade_handlers.go (hunk 23 above)
Comment (Missxiaoguo) [reply controllers/upgrade_handlers.go:86]: Please ignore this comment. The status should be reflected to the condition type.

File: controllers/upgrade_handlers.go (hunk 22 above)
Comment (Missxiaoguo) [reply controllers/upgrade_handlers.go]: Please ignore this comment. The status should be reflected to the condition type so it should be False

File: controllers/upgrade_handlers.go (hunk 21 above)
Comment (pixelsoccupied) [reply controllers/upgrade_handlers.go:61]: Updated

File: controllers/upgrade_handlers.go (hunk 23 above)
Comment (pixelsoccupied) [reply controllers/upgrade_handlers.go:89]: Updated

File: internal/backuprestore/backup.go (hunk 24 above)
Comment (pixelsoccupied) [reply internal/backuprestore/backup.go]: Removed the custom status and instead using velvero status.phase

File: controllers/upgrade_handlers.go (hunk 25 above)
Comment (pixelsoccupied) [reply controllers/upgrade_handlers.go]: Updated implementation to make it clearer

**Comment by pixelsoccupied** [comment]
We will need some tests for this as there some edge cases which we might miss in case minor/major refactoring. Maybe kuttle/e2e test will benefit from this? I'll do that with another PR though. wdyt @jc-rh @Missxiaoguo

**Comment by jc-rh** [comment]
> We will need some tests for this as there some edge cases which we might miss in case minor/major refactoring. Maybe kuttle/e2e test will benefit from this? I'll do that with another PR though. wdyt @jc-rh @Missxiaoguo

kuttl would be a bit difficult. Similar to TALM ones where we use scripts to update ACM CR status, we would need to fake OADP behavior. Might want to try UT first.
